syntax = "proto3";

package fintekkers.models.security.bond;

import "fintekkers/models/security/bond/auction_type.proto";

import "fintekkers/models/util/decimal_value.proto";
import "fintekkers/models/util/local_date.proto";
import "fintekkers/models/util/local_timestamp.proto";
// import "fintekkers/models/util/uuid.proto";

// import "fintekkers/models/security/identifier/identifier.proto";
// import "fintekkers/models/security/security_type.proto";
// import "fintekkers/models/security/security_quantity_type.proto";
// import "fintekkers/models/security/coupon_frequency.proto";
// import "fintekkers/models/security/coupon_type.proto";

option java_multiple_files = true;
option java_outer_classname = "IssuanceProtos";

/*
    Issuance proto contains the following:
        - changes in outstanding quantities.
            - auction
                - auction date
                - issue date (note: original issue date is found on the base security)
                - announcement date
                - pre-auction outstanding quantity
                - offering amount
                - auction type (e.g. single price)
                - price (if single price auction)
                - award details, i.e. below. The above should apply to corproate issuance also
<PrimaryDealerTendered>136650000000</PrimaryDealerTendered>
<PrimaryDealerAccepted>32879650000</PrimaryDealerAccepted>
<DirectBidderTendered>9525000000</DirectBidderTendered>
<DirectBidderAccepted>3548785000</DirectBidderAccepted>
<IndirectBidderTendered>9055736500</IndirectBidderTendered>
<IndirectBidderAccepted>8145454000</IndirectBidderAccepted>
<CompetitiveTendered>155230736500</CompetitiveTendered>
<CompetitiveAccepted>44573889000</CompetitiveAccepted>
<NonCompetitiveAccepted>326458500</NonCompetitiveAccepted>
<SOMATendered>0</SOMATendered>
<SOMAAccepted>0</SOMAAccepted>
<FIMATendered>100000000</FIMATendered>
<FIMAAccepted>100000000</FIMAAccepted>
<TotalTendered>155657195000</TotalTendered>
<TotalAccepted>45000347500</TotalAccepted>
            - tender offer
            - ?

    
*/
message IssuanceProto {
  string object_class = 1;
  string version = 2;

  //Primary Key not needed currently as these will be stored on the security object
//   fintekkers.models.util.UUIDProto uuid = 5;
  fintekkers.models.util.LocalTimestampProto as_of = 6;
//   bool is_link = 7;

  fintekkers.models.util.LocalTimestampProto valid_from = 8;
  fintekkers.models.util.LocalTimestampProto valid_to = 9;


  /**
  The date the auction was announced, has no bearing on the economics of the bond
   */
  fintekkers.models.util.LocalDateProto auction_announcement_date = 20;

  /**
  The date of the issue relating to this auction, i.e. when the securities will be issued to
  buyers
   */
  fintekkers.models.util.LocalDateProto auction_issue_date = 21;

  /**
  The amount of this security that is outstanding in the market before this auction
   */
  fintekkers.models.util.DecimalValueProto post_auction_outstanding_quantity = 22;

  /**
  The amount of bond issuance expected during this auction. This is an indicative amount and
  may not be the exact amount issued. For example in US Treasuries the amount issued to competitive
  bidders (e.g. primary dealers) might âˆšary from this number. Additional bonds may be bought by
  Federal programs such as SOMA/FIMA.
   */
  fintekkers.models.util.DecimalValueProto auction_offering_amount = 23;

  /***
  The type of auction, see the enum for details on types.
   */
  fintekkers.models.security.bond.AuctionTypeProto auction_type = 24;

  /**
  If the auction was a single price auction, this is the price paid by bidders
   */
  fintekkers.models.util.DecimalValueProto price_for_single_price_auction = 25;

  /**
  The total amount of the security sold by this auction.
   */
  fintekkers.models.util.DecimalValueProto total_accepted = 26;

  /**
  The amount of this security that will mature on the security's maturity date
   */
  fintekkers.models.util.DecimalValueProto mature_security_amount = 27;

}
